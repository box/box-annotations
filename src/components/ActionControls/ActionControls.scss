@import '../../variables';

.ba {
    .bcs-at-mention-tip {
        display: none;
    }

    .ba-action-controls {
        margin: 0;
        color: $fours;
        white-space: normal;
        background: white;
    }

    .ba-annotation-input-container {
        width: 252px; // Hard-coded width to solve layout issues
        border-top: none;
    }

    .bcs-comment .bcs-comment-headline {
        flex-flow: column wrap;
    }

    .bcs-comment .bcs-comment-content .bcs-comment-user-name ~ .bcs-comment-created-at {
        margin: 0;
    }

    .ba-annotator-label ~ .ba-action-controls .ba-action-controls-draw {
        margin-left: 5px;
    }
}
