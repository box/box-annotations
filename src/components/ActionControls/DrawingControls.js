// @flow
import React from 'react';

import PlainButton from 'box-react-ui/lib/components/plain-button';
import IconCheck from 'box-react-ui/lib/icons/general/IconCheck';
import IconTrash from 'box-react-ui/lib/icons/general/IconTrash';

import './DrawingControls.scss';

const CLASS_DRAW_CONTROLS = 'ba-action-controls-draw';
const CLASS_DRAWING_SAVE_BTN = 'ba-drawing-save-btn';
const CLASS_DRAWING_DELETE_BTN = 'ba-drawing-delete-btn';

type Props = {
    canDelete: boolean,
    isPending: boolean,
    onCreate: Function,
    onDelete: Function,
};

const DrawingControls = ({ isPending, canDelete, onCreate, onDelete }: Props) => (
    <div className={CLASS_DRAW_CONTROLS}>
        {isPending && (
            <PlainButton className={CLASS_DRAWING_SAVE_BTN} onClick={onCreate} type="button">
                <IconCheck />
            </PlainButton>
        )}
        {canDelete && (
            <PlainButton className={CLASS_DRAWING_DELETE_BTN} onClick={onDelete} type="button">
                <IconTrash height={20} width={20} />
            </PlainButton>
        )}
    </div>
);

DrawingControls.defaultProps = {
    canDelete: false,
};

export default DrawingControls;
