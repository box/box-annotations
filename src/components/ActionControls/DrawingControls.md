### Description
Component to represent the drawing action buttons in the annotation dialog

### Examples
```jsx
const DrawingControlsExamples = require('examples').DrawingControlsExamples;

<DrawingControlsExamples />
```
