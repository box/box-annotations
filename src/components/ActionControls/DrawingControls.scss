@import '../../variables';

.ba {
    .ba-action-controls-draw {
        display: flex;
    }

    .ba-action-controls-draw .btn-plain {
        flex-grow: 1;
        padding-left: 5px;
        padding-right: 5px;

        .fill-color {
            fill: $sevens;
        }
    }

    .icon-trash {
        margin-top: 4px;
    }

    .ba-action-controls-draw .btn-plain:hover {
        padding-left: 5px;
        padding-right: 5px;

        .fill-color {
            fill: $better-black;
        }
    }

    // MOBILE & TABLET CSS
    @media #{$mobile}, #{$tablet} {
        .ba-inline .ba-action-controls-draw {
            background: white;
            border-top: 1px solid $seesee;
            bottom: 0;
            box-shadow: 0 2px 6px 0 rgba(0, 0, 0, .1);
            left: 0;
            margin: 0;
            padding: 15px;
            position: fixed;
            width: 100%;
        }

        .ba-action-controls-draw .btn-plain,
        .ba-action-controls-draw .btn-plain:hover {
            padding: 10px;
        }

        .ba-drawing-save-btn,
        .ba-drawing-save-btn:hover {
            border-right: 1px solid $seesee;
        }
    }
}
