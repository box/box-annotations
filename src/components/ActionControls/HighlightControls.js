// @flow
import React from 'react';
import classNames from 'classnames';

import PlainButton from 'box-ui-elements/es/components/plain-button';
import IconHighlightAnnotation from 'box-ui-elements/es/icons/annotations/IconHighlightAnnotation';
import IconHighlightCommentAnnotation from 'box-ui-elements/es/icons/annotations/IconHighlightCommentAnnotation';

import './HighlightControls.scss';

const CLASS_HIGHLIGHT_CONTROLS = 'ba-action-controls-highlight';
const CLASS_HIGHLIGHT_BTN = 'ba-highlight-btn';
const CLASS_HIGHLIGHT_COMMENT_BTN = 'ba-highlight-comment-btn';
const CLASS_SAVED_HIGHLIGHT = 'ba-saved-highlight';

type Props = {
    canAnnotateAndDelete: boolean,
    canComment: boolean,
    isPending?: boolean,
    onCommentClick: Function,
    onCreate: Function,
    onDelete: Function,
};

const HighlightControls = ({
    canAnnotateAndDelete,
    canComment,
    onCreate,
    onDelete,
    onCommentClick,
    isPending,
}: Props) => (
    <div className={CLASS_HIGHLIGHT_CONTROLS}>
        {canAnnotateAndDelete && (
            <PlainButton className={CLASS_HIGHLIGHT_BTN} onClick={isPending ? onCreate : onDelete} type="button">
                <IconHighlightAnnotation
                    className={classNames({
                        [CLASS_SAVED_HIGHLIGHT]: !isPending,
                    })}
                    height={20}
                    width={20}
                />
            </PlainButton>
        )}
        {canComment && (
            <PlainButton className={CLASS_HIGHLIGHT_COMMENT_BTN} onClick={onCommentClick} type="button">
                <IconHighlightCommentAnnotation height={20} width={20} />
            </PlainButton>
        )}
    </div>
);

HighlightControls.defaultProps = {
    canAnnotateAndDelete: false,
    isPending: false,
};

export default HighlightControls;
