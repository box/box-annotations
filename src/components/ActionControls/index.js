// @flow
export { default } from './ActionControls';
