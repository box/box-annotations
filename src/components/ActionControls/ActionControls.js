// @flow
import * as React from 'react';
import noop from 'lodash/noop';

import ApprovalCommentForm from '../../../third-party/components/ApprovalCommentForm';

import { TYPES } from '../../constants';
import HighlightControls from './HighlightControls';
import DrawingControls from './DrawingControls';

import './ActionControls.scss';

const CLASS_ACTION_CONTROLS = 'ba-action-controls';
const CLASS_INPUT_CONTAINER = 'ba-annotation-input-container';

type Props = {
    canComment: boolean,
    canDelete: boolean,
    hasComments: boolean,
    id?: string,
    isPending: boolean,
    onCancel: Function,
    onCommentClick: Function,
    onCreate: Function,
    onDelete: Function,
    type: AnnotationType,
};

type State = {
    isInputOpen: boolean,
};

const NULL_USER = {};

class ActionControls extends React.Component<Props, State> {
    static defaultProps = {
        canComment: false,
        canDelete: false,
        isPending: false,
    };

    state = {
        isInputOpen: true,
    };

    componentDidMount() {
        this.onFocus();
    }

    componentWillUnmount() {
        this.setState({ isInputOpen: false });
    }

    onFocus = () => {
        this.setState({ isInputOpen: true });
    };

    onCancel = (event: Event) => {
        event.stopPropagation();
        event.preventDefault();
        this.setState({ isInputOpen: false });

        const { onCancel } = this.props;
        onCancel();
    };

    onCreate = ({ text }: { text: string }) => {
        const { onCreate, type } = this.props;
        onCreate(type, text);
        this.setState({ isInputOpen: true });
    };

    onDelete = (event: Event) => {
        event.stopPropagation();
        event.preventDefault();
        this.setState({ isInputOpen: false });

        const { onDelete, id } = this.props;
        onDelete({ id });
    };

    determineControls(type: string): ?React.Node {
        const { canComment, canDelete, onCommentClick, isPending, hasComments } = this.props;
        const { isInputOpen } = this.state;

        // NOTE: ActionControls are only displayed when the user has canAnnotate permissions
        // Any annotation created by the current user will also be deletable by that user
        switch (type) {
            case TYPES.highlight:
                // Do not display plain highlight controls if the user does not have
                // permissions to create/delete/comment annotations
                if (!canDelete && !canComment) {
                    return null;
                }

                return (
                    <HighlightControls
                        canAnnotateAndDelete={canDelete}
                        canComment={canComment}
                        isPending={isPending}
                        onCommentClick={onCommentClick}
                        onCreate={this.onCreate}
                        onDelete={this.onDelete}
                    />
                );
            case TYPES.highlight_comment:
                // Display the ApprovalCommentForm when adding the first comment
                // to a plain highlight or additional comments to highlight
                // comment annotations
                if (isPending || hasComments) {
                    return (
                        <ApprovalCommentForm
                            className={CLASS_INPUT_CONTAINER}
                            // $FlowFixMe
                            createComment={this.onCreate}
                            // $FlowFixMe
                            getAvatarUrl={noop}
                            isEditing
                            isOpen={isInputOpen}
                            onCancel={this.onCancel}
                            onFocus={this.onFocus}
                            onSubmit={noop}
                            // $FlowFixMe
                            user={NULL_USER}
                        />
                    );
                }

                return (
                    <HighlightControls
                        canAnnotateAndDelete={canDelete}
                        canComment
                        isPending={isPending}
                        onCommentClick={onCommentClick}
                        onCreate={this.onCreate}
                        onDelete={this.onDelete}
                    />
                );
            case TYPES.draw:
                // User cannot see any actions if they are unable to delete existing drawings.
                // Users will have delete permissions on any drawings that they create
                if (!canDelete) {
                    return null;
                }

                return (
                    <DrawingControls
                        canDelete={canDelete}
                        isPending={isPending}
                        onCreate={this.onCreate}
                        onDelete={this.onDelete}
                    />
                );
            case TYPES.point:
                // Users will always be able to add comments to existing point annotations
                // irrespective of whether or not they can create new point annotations
                return (
                    <ApprovalCommentForm
                        className={CLASS_INPUT_CONTAINER}
                        // $FlowFixMe
                        createComment={this.onCreate}
                        // $FlowFixMe
                        getAvatarUrl={noop}
                        isEditing
                        isOpen={isInputOpen}
                        onCancel={this.onCancel}
                        onFocus={this.onFocus}
                        onSubmit={noop}
                        // $FlowFixMe
                        user={NULL_USER}
                    />
                );
            default:
                return null;
        }
    }

    render() {
        const { type } = this.props;
        return <div className={CLASS_ACTION_CONTROLS}>{this.determineControls(type)}</div>;
    }
}

export default ActionControls;
