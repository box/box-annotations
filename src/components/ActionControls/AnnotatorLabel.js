// @flow
import React from 'react';
import { injectIntl } from 'react-intl';

import CommentText from '../../../third-party/components/Comment/CommentText';

import { TYPES } from '../../constants';
import { isHighlightAnnotation } from '../../util';

import messages from './messages';

import './ActionControls.scss';

type Props = {
    id: string,
    type: AnnotationType,
    currentUser?: User,
    intl: any
};

class AnnotatorLabel extends React.PureComponent<Props> {
    getAnnotationLabel() {
        const { type, currentUser, intl } = this.props;
        const anonymousUserName = intl.formatMessage(messages.anonymousUserName);
        const annotatorName = !!currentUser && !!currentUser.name ? currentUser.name : anonymousUserName;

        if (isHighlightAnnotation(type)) {
            return intl.formatMessage(messages.whoHighlighted, { name: annotatorName });
        }

        if (type === TYPES.draw) {
            return intl.formatMessage(messages.whoDrew, { name: annotatorName });
        }

        return intl.formatMessage(messages.whoAnnotated, { name: annotatorName });
    }

    render() {
        const { id } = this.props;
        const annotatorLabelMessage = this.getAnnotationLabel();

        return (
            <span className='ba-annotation-label'>
                <CommentText id={id} tagged_message={annotatorLabelMessage} translationEnabled={false} />
            </span>
        );
    }
}

export default injectIntl(AnnotatorLabel);
