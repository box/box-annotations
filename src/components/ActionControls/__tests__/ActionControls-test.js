import * as React from 'react';
import { shallow } from 'enzyme';

import ActionControls from '../ActionControls';

const onCreate = jest.fn();
const onCommentClick = jest.fn();
const onCancel = jest.fn();
const onDelete = jest.fn();

const event = {
    stopPropagation: jest.fn(),
    preventDefault: jest.fn(),
};

describe('components/ActionControls', () => {
    const render = (props = {}) =>
        shallow(
            <ActionControls
                canDelete={false}
                hasComments={false}
                isPending={false}
                onCancel={onCancel}
                onCommentClick={onCommentClick}
                onCreate={onCreate}
                onDelete={onDelete}
                {...props}
            />,
        );

    test('should render no controls if the type does not match a valid type', () => {
        const wrapper = render({ type: 'invalid' });
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('HighlightControls').exists()).toBeFalsy();
        expect(wrapper.find('DrawingControls').exists()).toBeFalsy();
        expect(wrapper.find('ApprovalCommentForm').exists()).toBeFalsy();
    });

    test('should render no controls if user cannot delete or comment on a plain highlight annotation', () => {
        const wrapper = render({ type: 'highlight' });
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('HighlightControls').exists()).toBeFalsy();
        expect(wrapper.find('DrawingControls').exists()).toBeFalsy();
        expect(wrapper.find('ApprovalCommentForm').exists()).toBeFalsy();
    });

    test('should correctly render controls for plain highlight annotations without delete permissions', () => {
        const wrapper = render({ type: 'highlight', canComment: true });
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('HighlightControls').exists()).toBeTruthy();
    });

    test('should correctly render controls for plain highlight annotations without comment permissions', () => {
        const wrapper = render({ type: 'highlight', canDelete: true });
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('HighlightControls').exists()).toBeTruthy();
    });

    test('should correctly render controls for plain highlight annotations', () => {
        const wrapper = render({ canDelete: true, type: 'highlight' });
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('HighlightControls').exists()).toBeTruthy();
    });

    test('should correctly render controls for pending highlight comment annotations', () => {
        const wrapper = render({ isPending: true, type: 'highlight-comment' });
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('ApprovalCommentForm').exists()).toBeTruthy();
    });

    test('should correctly render controls for highlight comment annotations without any existing comments', () => {
        const wrapper = render({ type: 'highlight-comment', hasComments: true });
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('ApprovalCommentForm').exists()).toBeTruthy();
    });

    test('should correctly render controls for highlight comment annotations', () => {
        const wrapper = render({ type: 'highlight-comment' });
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('HighlightControls').exists()).toBeTruthy();
    });

    test('should render no controls if user cannot delete or comment on a plain highlight annotation', () => {
        const wrapper = render({ type: 'draw' });
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('HighlightControls').exists()).toBeFalsy();
        expect(wrapper.find('DrawingControls').exists()).toBeFalsy();
        expect(wrapper.find('ApprovalCommentForm').exists()).toBeFalsy();
    });

    test('should correctly render controls for drawing annotations', () => {
        const wrapper = render({ type: 'draw', canDelete: true });
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('DrawingControls').exists()).toBeTruthy();
    });

    test('should correctly render controls for point annotations', () => {
        const wrapper = render({ type: 'point' });
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('ApprovalCommentForm').exists()).toBeTruthy();
    });

    test('should update state and call onDelete() prop on delete button click', () => {
        const wrapper = render({ canDelete: true, type: 'highlight' });
        const instance = wrapper.instance();

        instance.onDelete(event);
        wrapper.update();

        expect(event.stopPropagation).toBeCalled();
        expect(event.preventDefault).toBeCalled();
        expect(onDelete).toBeCalled();
        expect(wrapper.state().isInputOpen).toEqual(false);
    });

    test('should update state and call onCreate() prop on create button click', () => {
        const wrapper = render({ canDelete: true, type: 'highlight' });
        const instance = wrapper.instance();

        instance.onCreate(event);
        wrapper.update();

        expect(onCreate).toBeCalled();
        expect(wrapper.state().isInputOpen).toEqual(true);
    });

    test('should update state and call onCancel() prop on cancel button click', () => {
        const wrapper = render({ canDelete: true, type: 'highlight' });
        const instance = wrapper.instance();

        instance.onCancel(event);
        wrapper.update();

        expect(event.stopPropagation).toBeCalled();
        expect(event.preventDefault).toBeCalled();
        expect(onCancel).toBeCalled();
        expect(wrapper.state().isInputOpen).toEqual(false);
    });

    test('should update state to open the input onFocus()', () => {
        const wrapper = render({ canDelete: true, type: 'highlight' });
        const instance = wrapper.instance();

        instance.onFocus(event);
        wrapper.update();

        expect(wrapper.state().isInputOpen).toEqual(true);
    });

    test('should update state to close the input componentWillUnmount()', () => {
        const wrapper = render({ canDelete: true, type: 'highlight' });
        const instance = wrapper.instance();

        instance.componentWillUnmount(event);
        wrapper.update();

        expect(wrapper.state().isInputOpen).toEqual(false);
    });

    test('should focus the input on componentDidMount()', () => {
        const wrapper = render({ canDelete: true, type: 'highlight' });
        expect(wrapper.state().isInputOpen).toEqual(true);
    });
});
