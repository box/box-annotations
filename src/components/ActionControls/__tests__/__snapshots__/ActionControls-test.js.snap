// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ActionControls should correctly render controls for drawing annotations 1`] = `
<div
  className="ba-action-controls"
/>
`;

exports[`components/ActionControls should correctly render controls for highlight comment annotations 1`] = `
<div
  className="ba-action-controls"
>
  <HighlightControls
    canAnnotateAndDelete={true}
    canComment={true}
    isPending={false}
    onCommentClick={[MockFunction]}
    onCreate={[Function]}
    onDelete={[Function]}
  />
</div>
`;

exports[`components/ActionControls should correctly render controls for highlight comment annotations without delete permissions 1`] = `
<div
  className="ba-action-controls"
>
  <HighlightControls
    canAnnotateAndDelete={false}
    canComment={true}
    isPending={false}
    onCommentClick={[MockFunction]}
    onCreate={[Function]}
    onDelete={[Function]}
  />
</div>
`;

exports[`components/ActionControls should correctly render controls for pending highlight comment annotations 1`] = `
<div
  className="ba-action-controls"
>
  <ApprovalCommentForm
    className="ba-annotation-input-container"
    createComment={[Function]}
    getAvatarUrl={[Function]}
    isEditing={true}
    isOpen={true}
    onCancel={[Function]}
    onFocus={[Function]}
    onSubmit={[Function]}
    user={Object {}}
  />
</div>
`;

exports[`components/ActionControls should correctly render controls for plain highlight annotations 1`] = `
<div
  className="ba-action-controls"
>
  <HighlightControls
    canAnnotateAndDelete={true}
    canComment={false}
    isPending={false}
    onCommentClick={[MockFunction]}
    onCreate={[Function]}
    onDelete={[Function]}
  />
</div>
`;

exports[`components/ActionControls should correctly render controls for plain highlight annotations without delete permissions 1`] = `
<div
  className="ba-action-controls"
/>
`;

exports[`components/ActionControls should correctly render controls for point annotations 1`] = `
<div
  className="ba-action-controls"
>
  <ApprovalCommentForm
    className="ba-annotation-input-container"
    createComment={[Function]}
    getAvatarUrl={[Function]}
    isEditing={true}
    isOpen={true}
    onCancel={[Function]}
    onFocus={[Function]}
    onSubmit={[Function]}
    user={Object {}}
  />
</div>
`;
